In cron, access denied to big query should be an error

Thank you for contributing to this project!

- Make sure to search the [issues](https://github.com/tl-its-umich-edu/my-learning-analytics/issues) for duplicates first!

## Expected behavior (A clear and concise description of what you expected to happen) : 
If you cannot connect to big query it prints up a message

`[2019-08-16 16:14:20 -0400] [1] [INFO] 403 POST https://www.googleapis.com/bigquery/v2/projects/xxx-xxx-xxx/jobs: Access Denied: Project xxx-xxx-xxx: The user xxx-xxx-xxx@xxx-xxx-xxx.iam.gserviceaccount.com does not have bigquery.jobs.create permission in project xxx-xxx-xxx.`

## Describe the bug (Tell us what happens instead of the expected behavior) :
Instead of an [INFO] This really should probably be an [ERROR] message which would fail the job

--- a/dashboard/cron.py
+++ b/dashboard/cron.py
@@ -296,7 +296,7 @@ class DashboardCronJob(CronJobBase):
                 total_bytes_billed += bq_job.total_bytes_billed
             # If there is an exception here print an error message and return
             except Exception as e:
-                err_msg = f"There was an exception running the bigquery job of {e.message}"
+                err_msg = f"There was an exception running the BigQuery job of {e.message}"
                 # Print this as an error so it will trigger an error in the cron
                 logger.error(err_msg)
                 return status + err_msg

--- a/dashboard/cron.py
+++ b/dashboard/cron.py
@@ -289,10 +289,17 @@ class DashboardCronJob(CronJobBase):
             job_config.query_parameters = query_params
 
             # Location must match that of the dataset(s) referenced in the query.
-            bq_query = bigquery_client.query(final_bq_query, location='US', job_config=job_config)
-
-            resource_access_df: pd.DataFrame = bq_query.to_dataframe()
-            total_bytes_billed += bq_query.total_bytes_billed
+            try:
+                bq_job = bigquery_client.query(final_bq_query, location='US', job_config=job_config)
+                # This is the call that could result in an exception
+                resource_access_df: pd.DataFrame = bq_job.result().to_dataframe()
+                total_bytes_billed += bq_job.total_bytes_billed
+            # If there is an exception here print an error message and return
+            except Exception as e:
+                err_msg = f"There was an exception running the bigquery job of {e.message}"
+                # Print this as an error so it will trigger an error in the cron
+                logger.error(err_msg)
+                return status + err_msg
 
             resource_access_row_count = len(resource_access_df)
             if resource_access_row_count == 0:

--- a/dashboard/cron.py
+++ b/dashboard/cron.py
@@ -289,17 +289,10 @@ class DashboardCronJob(CronJobBase):
             job_config.query_parameters = query_params
 
             # Location must match that of the dataset(s) referenced in the query.
-            try:
-                bq_job = bigquery_client.query(final_bq_query, location='US', job_config=job_config)
-                # This is the call that could result in an exception
-                resource_access_df: pd.DataFrame = bq_job.result().to_dataframe()
-                total_bytes_billed += bq_job.total_bytes_billed
-            # If there is an exception here print an error message and return
-            except Exception as e:
-                err_msg = f"There was an exception running the BigQuery job of {e.message}"
-                # Print this as an error so it will trigger an error in the cron
-                logger.error(err_msg)
-                return status + err_msg
+            bq_job = bigquery_client.query(final_bq_query, location='US', job_config=job_config)
+            # This is the call that could result in an exception
+            resource_access_df: pd.DataFrame = bq_job.result().to_dataframe()
+            total_bytes_billed += bq_job.total_bytes_billed
 
             resource_access_row_count = len(resource_access_df)
             if resource_access_row_count == 0:
@@ -605,8 +598,8 @@ class DashboardCronJob(CronJobBase):
 
         status = ""
 
-        run_start = datetime.now()
-        status += f"Start cron: {str(run_start)}\n"
+        run_start = datetime.now(pytz.UTC)
+        status += f"Start cron: {str(run_start)} UTC\n"
 
         course_verification = self.verify_course_ids()
         invalid_course_id_list = course_verification.invalid_course_ids
@@ -631,6 +624,8 @@ class DashboardCronJob(CronJobBase):
             logger.info("Skipping course-related table updates...")
             status += "Skipped course-related table updates.\n"
         else:
+            # Update the date unless there is an exception
+            exception_in_run = False
             logger.info("** course")
             status += self.update_course(course_verification.course_data)
 
@@ -649,7 +644,9 @@ class DashboardCronJob(CronJobBase):
                     status += self.update_with_bq_access()
                     status += self.update_canvas_resource()
                 except Exception as e:
-                    logger.exception("Exception running BigQuery update")
+                    logger.error(f"Exception running BigQuery update: {str(e)}")
+                    status += str(e)
+                    exception_in_run = True
 
         if settings.DATA_WAREHOUSE_IS_UNIZIN:
             logger.info("** informational")
@@ -661,7 +658,11 @@ class DashboardCronJob(CronJobBase):
             logger.warning(f'No data was pulled for these courses.')
         
         # Set all of the courses to have been updated now (this is the same set update_course runs on)
-        Course.objects.filter(id__in=self.valid_locked_course_ids).update(data_last_updated=datetime.now(pytz.UTC))
+        if not exception_in_run:
+            logger.info(f"Updating all valid courses from when this run was started at {run_start}")
+            Course.objects.filter(id__in=self.valid_locked_course_ids).update(data_last_updated=run_start)
+        else:
+            logger.warn("data_last_updated not updated because of an Exception during this run")
 
         status += "End cron: " +  str(datetime.now()) + "\n"
 

--- a/requirements.txt
+++ b/requirements.txt
@@ -39,7 +39,7 @@ pysaml2==5.1.0
 SQLAlchemy==1.3.18
 psycopg2==2.8.5
 mysqlclient==1.4.6
-google-cloud-bigquery==1.26.0
+google-cloud-bigquery[pandas]==2.4.0
 
 pinax-eventlog==2.0.3
 debugpy==1.2.0

